src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/InteropHelpers.cs

```diff
@@ -598,6 +599,43 @@ internal static unsafe void CleanupVariant(IntPtr pDstNativeVariant)
 #endif
         }
 
+        public static unsafe ICustomMarshaler InitializeCustomMarshaller(RuntimeTypeHandle pParameterType, Type pMarshallerType, string cookie, delegate*<string, object> getInstanceMethod)
+        {
+            if (pMarshallerType == null)
+            {
+                throw new TypeLoadException();
+            }
+
+            if (pMarshallerType.IsGenericTypeDefinition)
+            {
+                throw new TypeLoadException();
+            }
+
+            if (!pMarshallerType.IsAssignableTo(typeof(ICustomMarshaler)))
```

Comment:
This is making interop a lot slower just to throw a compatible exception. It might not be the right compromise.

(Also grabbing a System.Type is not free either.)

We have been compromising on throwing the right exception type if it's more than a couple lines of code, or comes with size/perf issues.

Reply:
Okay, I make that change to be pedantic about passing tests without changing `ExceptionStringID` enum, so I will revert `Type` usage, and will use `RuntimeTypeHandle` again with wrong exception temporary.

Regarding field. Let's take `String` as type which would be marshalled.
There 3 cases
1. String as parameter. This works now and CoreCLR
https://github.com/dotnet/runtime/blob/c2b7638a5b6f3a6a52540244f3f85fba02284459/src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs#L25

2. String as field inside class. This works now and in CoreCLR
https://github.com/dotnet/runtime/blob/c2b7638a5b6f3a6a52540244f3f85fba02284459/src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs#L96

3. String as field inside structure. This presumable works now, but should not in CoreCLR
https://github.com/dotnet/runtime/blob/c2b7638a5b6f3a6a52540244f3f85fba02284459/src/tests/Interop/ICustomMarshaler/Primitives/ICustomMarshaler.cs#L554

Cases 2 and 3 identical except that field inside class or struct. I need something more then `isField`. Something like `isStruct`. I maybe miss something obvious?

Reply:
Maybe I'm missing something. Case 3 attempts to marshal following structure:

```csharp
        public struct StructWithCustomMarshalerField
        {
            [MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(StringForwardingCustomMarshaler))]
            public string Field;
        }

// With this P.invoke:

public static extern int StructWithCustomMarshalerFieldMethod(StructWithCustomMarshalerField c);
```

This fails because the field is decorated as CustomMarshaler. It's what `isField` covers.

Case 2 marshals this class:

```csharp

        public class StringContainer
        {
            public string Value { get; set; }
        }

// With this pinvoke

        public static extern StringContainer MarshalerOnClassTypeMethod([MarshalAs(UnmanagedType.CustomMarshaler, MarshalTypeRef = typeof(ClassForwardingCustomMarshaler))] StringContainer str);
```

This works because CustomMarshaller is on a parameter, not on a field. The `isField` check would not be hit for this case because the MarshalAs is not on the field.

Reply:
Aah, Yes. I definitely overlook tha. Thanks.